### ISSUE TITLE
Raw Lossles Raster reading support for JPEGImageReader

### ISSUE THREAD

**Issue by xiexed** [haraldk/TwelveMonkeys#234]
Hello. [JPEGImageReader.java#L929](https://github.com/haraldk/TwelveMonkeys/blob/master/imageio/imageio-jpeg/src/main/java/com/twelvemonkeys/imageio/plugins/jpeg/JPEGImageReader.java#L929) at this moment does not support reading **raw** `Raster` data. Because the `readRaster` method always delegates to system reader.

This is actual for me in view of Lossless JPEG support (pull request #215), but maybe it also should be supported in other codecs.

At this moment I've fixed it this way:

```java
    @Override
    public Raster readRaster(int imageIndex, ImageReadParam param) throws IOException {

        if(getSOF().marker == JPEG.SOF3){
            JPEGLosslessDecoder decoder = new JPEGLosslessDecoder(getBytes((ImageInputStream) this.getInput()));
            int[][] decoded = decoder.decode();
            switch(decoder.getPrecision())
            {
                case 8: {
                    DataBufferByte buff = new DataBufferByte(decoded[0].length, 1);
                    for (int i = 0; i < decoded[0].length; i++) {
                        buff.getData()[i] = (byte) decoded[0][i];
                    }
                    return Raster.createRaster(new BandedSampleModel(DataBuffer.TYPE_BYTE, decoder.getDimX(), decoder.getDimY(), 1),
                            buff, null);
                }
                case 16: {
                    DataBufferUShort buff = new DataBufferUShort(decoded[0].length, 1);
                    for (int i = 0; i < decoded[0].length; i++) {
                        buff.getData()[i] = (short) decoded[0][i];
                    }
                    return Raster.createRaster(new BandedSampleModel(DataBuffer.TYPE_USHORT, decoder.getDimX(), decoder.getDimY(), 1),
                            buff, null);
                }
                default:
                    throw new IOException("JPEG Lossless with " + decoder.getPrecision() + " bit precision and 1 component cannot be decoded");
            }

        }

        return delegate.readRaster(imageIndex, param);
    }
```

This works for me (after applying #215 of course), but I am not experienced in internals of Java imaging, so maybe there is a more convenient way. And also, possibly this code better should be putted into ongoing [JPEGLosslessDecoderWrapper](https://github.com/HermannKroll/TwelveMonkeys/blob/94a018f/imageio/imageio-jpeg/src/main/java/com/twelvemonkeys/imageio/plugins/jpeg/lossless/JPEGLosslessDecoderWrapper.java) class of pull request #215

Thanks

**Comment by haraldk** [comment]
Thanks for reporting!

Do you also have a (preferably small) test file to go along with the issue, so that we can create a test case for it (to avoid breaking this feature again in the future).

-- 
Harald K

**Comment by xiexed** [comment]
Oh... it will not be very easy. All the database that we use is prohibited to publish and also it is needed to wrap out all DICOM staff to avoid bringing them into tests. But I will try to do something.

**Comment by HermannKroll** [comment]
Hi,

i have an JPEG Lossless test file. It's extracted out of an DICOM file and only contains the JPEG data inside the DICOM.

https://github.com/rii-mango/JPEGLosslessDecoder/files/61515/test0.pixeldata.zip

Best regards,
Hermann K

**Comment by haraldk** [comment]
@HermannKroll Thanks! 

I think that will work fine. Sorry about the slow progress on this issue and the lossless JPEG in general. I've been swamped up in other work lately, and need some time to clean up after the merge. But it will happen! :-)

-- 
Harald K

**Comment by haraldk** [comment]
Fixed as part of #182 

-- 
Harald K

**Comment by haraldk** [comment]
@xiexed This bugfix is now included in the official 3.3 release.

**Comment by xiexed** [comment]
Thats great! Thank you.
